Задание 7. Добавлен функционал открытия\создания счета.

gudkxv opened the pull request:
### Всем привет, посылаю новый PR - добавление функционала создаения счета.

**Добавлено:**

- Новый bean - CreateAccountCommand, где происходит обработка команды "/createAccount".
- MiddleResponseHandler - компонент, в котором обрабатывается HttpStatus ответа от middle. В зависимости от значения устанавливается текст возвращаемого сообщения. Помимо этого, внутри "костыльно" обрабатываются FeignException.

**Пример flow:**
204
![image](https://github.com/gpb-it-factory/gudkov-telegram-bot/assets/91263453/27cf7737-7fb4-4bec-935d-62196dce47aa)

![image](https://github.com/gpb-it-factory/gudkov-telegram-bot/assets/91263453/782ce08a-b690-484d-97c6-ae96d74fa5b7)

409
![image](https://github.com/gpb-it-factory/gudkov-telegram-bot/assets/91263453/e35961de-f469-42a5-a790-37ad2ac5431c)

![image](https://github.com/gpb-it-factory/gudkov-telegram-bot/assets/91263453/0206a993-56f5-48ba-bd4e-31c9fb8f299e)

!204 && !409

![image](https://github.com/gpb-it-factory/gudkov-telegram-bot/assets/91263453/26ae64a7-eef9-4447-8ae0-1d6f996f9d9a)

![image](https://github.com/gpb-it-factory/gudkov-telegram-bot/assets/91263453/711af32a-be58-4f31-bfad-d09ad212c5b6)

Commit:
Задание 7. Добавлен функционал открытия\создания счета. Немного переработана архитектура.

gudkxv commented:
И у меня возник небольшой вопрос, относительно валидации входных данных, мы же их на front - слое только принимаем, а валидируем (проверяем, что зарегистрирован) уже в middle?

NPriimak commented on src/main/java/gudkov/miit/tgbot/Commands/CreateAccountCommand.java:
```diff
@@ -0,0 +1,61 @@
+            CreateAccountRequestV2 createAccountRequestV2 = new CreateAccountRequestV2("My first and only awesome account");
+            replyMessage = responseHandler.handleBackendCreateAccountResponse(middleServiceApi.createUserAccountV2(userId, createAccountRequestV2),replyMessage);
+        } catch (FeignException feignException){
+            replyMessage = responseHandler.handleFeignExceptionAccountCreation(feignException,replyMessage);
```
Вот так делать не надо) 

Вообще никогда-никогда не меняйте ссылки в параметрах)  Это неочевидно, противоречит концепции чистых функций и сложно тестируется в больших приложениях. 

В данном случае все намного проще - если текст ответа возвращается неким responseHandler - то он и должен возвращаться строку, а не принимать в себя объект, модифицировать его и тд. Лучше тогда уже вернуть из этого метода строку, и спокойно проставить результат в основном методе. 

Плюс у нас метод называется - попробуйт сделать запрос для счета - а по факту он: 
1) Делает запрос
2) Вызывает сторонний сервис и в зависимости от ответа отдает строку
3) Модифицирует передаваемый параметр

Я бы сделал так: 

В основном методе споконо вызываем миддл слой.
Если получаем ошибку - в отдельном методе / сервисе спокойно парсим ошибку и получаем текст ответа, если все ок - получаем положительный ответ.
Формируем сообщение для пользователя - возвращаем. 

Что скажешь?

gudkxv replied:
Да, согласен, в одну функцию и правда многое накидал. Постарался исправить.

NPriimak commented on src/main/java/gudkov/miit/tgbot/Handlers/MiddleResponseHandler.java:
```diff
@@ -0,0 +1,56 @@
+ * Inside there is also functions that parse FeignException status code to change replyMessage text based on value;
+ */
+@Component
+public class MiddleResponseHandler {
```
На самом деле все это проще вытащить в класс клиент для миддл слоя и там обрабатывать ошибки, так как стоит только нам сменить реализацию с рест на очереди - все это становится бесполезным и нам прийдется переделывать всю бизнес логику

gudkxv replied:
После исправлений данный компонент ушел в небытие.

NPriimak reviewed (commented):
Все ок, хорошая работа! Как по мне допустили небольшую путаницу в логику отправки запросов - постарался расписать. Прошу посмотреть

Commit:
Reworked CreateAccountCommand main function.
Reworked RegisterCommand main function same style as CreateAccount.
Also removed API versioning and changed API generating task based on new API

gudkxv commented:
С замечаниями полностью согласен, вроде как подправил. Во время исправления оказалось, что компонент MiddleResponseHandler оказался не нужным, т.к теперь все ответы от мидла обрабатываются внутри бина, ответственного за обработку команды.

Сделано:

- Переработал CreateAccountCommand.  Теперь она, вроде как, работает, как подсказали. Убрал ненужный метод, кода поменьше, да и выглядеть стало получше.
- По той-же логике переработал RegistrationCommand. 
- После этого компонент MiddleResponseHandler оказался не испольхуемым, по сей причине был удален.
